### ISSUE TITLE
[65] data2vec: A General Framework for Self-supervised Learning in Speech, Vision and Language

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#93]
[paper](https://scontent-ssn1-1.xx.fbcdn.net/v/t39.8562-6/271974914_483120576492438_4239522333319653600_n.pdf?_nc_cat=107&ccb=1-5&_nc_sid=ae5e01&_nc_ohc=4-cMR5tUq4QAX-P3LNk&_nc_ht=scontent-ssn1-1.xx&oh=00_AT8YsMgdJQi2FjSwl85tjXB9bJaW1oEq5xWupPhghxo5pQ&oe=61F1FD91)
[code](https://github.com/pytorch/fairseq/tree/main/examples/data2vec)
[blog](https://ai.facebook.com/blog/the-first-high-performance-self-supervised-algorithm-that-works-for-speech-vision-and-text/)

wav2vec 2.0 저자의 논문.
speech, vision, language 를 모두 mask prediction 을 하는 task 로 pretraining 하는 논문
speech 든 vision 이든 language 든 모두 Mask prediction이 대세인 상황에서 충분히 나올만 했던 논문으로 보인다.
논문은 gopher 처럼 arxiv 에 올리지도 않았다 ㅋㅋ

![image](https://user-images.githubusercontent.com/16400591/150674933-af0aba41-d038-4675-9f85-d444a5b4d092.png)

방법은 엄청 단순하다.
EMA로 업데이트 되는 teacher 에 원본 이미지를 넣고 feature 를 뽑아놓는다. (target)
student 에는 mask 된 input 을 집어넣고,
끝에 있는 K layer 에서 feature level 로 loss 를 준다.

## Input process
- Vision
    - ViT-strategy 를 그대로. 16x16 input
    - masking 은 BeiT 처럼 blockwise
    - 16 patch masking with random aspect ratio
    - masking 은 60% 가 좋았음
- Speech
    - wav2vec 2.0 그대로. multi-layer 1-D CNN 통과 (16 kHz waveform to 50 Hz)
    - masking 도 wav2vec 2.0 처럼 span mask.
    - starting indices and mask the subsequent ten time-steps p = 0.065 확률로 masking. 
    - 결과적으로 49% 정도가 masking 되어짐.
- Language
    - BERT 그대로. sub-word units 이용.
    - masking 도 BERT 방식대로. 
    - 15% masking, 그 중 80% 는 learned mask token, 10%는 안 바꾸고, 10% 는 random vocab 으로.

## Teacher parameterization
∆ ← τ∆ + (1 − τ ) θ
τ 값은 linear 하게 증가시키도록 schedule. (from τ0 to τe)

## Targets
block l, time-step t  의 output 을 <img src="https://render.githubusercontent.com/render/math?math=a^l_t"> 라 하자.
이걸 normalize 해주면, <img src="https://render.githubusercontent.com/render/math?math=\hat{a}^l_t"> 
- speech 쪽은 layer norm
- vision, language 쪽은 parameter-less layer norm

time-step t 에서 target y 는 다음과 같다
![image](https://user-images.githubusercontent.com/16400591/150677383-87bda1f0-fb4f-4947-aa54-b4417f9c4be9.png)

## Loss
loss 는 smooth l1 loss
![image](https://user-images.githubusercontent.com/16400591/150677438-c350ced8-8b44-448e-a514-f33d98c75a08.png)

# Results
## Computer vision
괄호 수치는 각각 ViT-B, ViT-L
800 epoch pretrain (2,048 batch, 8,192 batch)
cosine LR (warmup 40 epoch, 80 epoch) 
Adam (lr_40 = 0.002,  0.001)
β = 2, K = 6 and τ = 0.9998
stochastic depth 0.2
![image](https://user-images.githubusercontent.com/16400591/150677973-f1c5a4e7-87eb-404d-899f-eb95d5f1cbca.png)

## Speech processing
Librispeech (960 hours data) pretraining (LS-960)
τ0 = 0.999, τe = 0.9999 and τn = 30, 000
![image](https://user-images.githubusercontent.com/16400591/150677522-44a4b599-e0e9-4404-b781-09732f1dda0e.png)

##  Natural language processing
BERT 랑 맞춤.
Books Corpus, English Wikipedia 에 256 batch 로 1M 번 update
Baseline ==> roberta
τ0 = 0.999, τe = 0.9999
τn = 100, 000
β = 4, K = 10
![image](https://user-images.githubusercontent.com/16400591/150677685-320bf645-a901-4598-858d-34411ae13d52.png)

### K 에 따른 성능
![image](https://user-images.githubusercontent.com/16400591/150678434-e7aa46cc-d830-48bd-9bfd-b5165d7e1a3c.png)
K=1 이면 BYOL 임.

### 어디서 feature 를 뽑을까
![image](https://user-images.githubusercontent.com/16400591/150678516-c4eb8392-88d8-47b9-8374-f498c089798c.png)

FFN 이 왜 잘 될까? 저자들은 self-attention output 이 residual connection 전에 있고, 그 때의 features 들은 다른 time-steps 에 bias 되기 때문이라 한다.

## Discussion
자체적으로 target 생성 후 학습시키는 알고리즘은 보통 representation collapse가 일어남.
모델이 모든 masking segment들에 대해 유사한 representation을 생성할 때 발생함.

이 문제를 해결하기 위해 다양한 전략이 제안되어 왔음.
wav2vec 2.0은 positive sample 과 negative sample 에서 같은 target representation을 사용. 
BYOL은 teacher 쪽은 loss minimization 을 하지 않음. 
VicReg는 explicit loss를 추가해서 representation 간의 분산을 높임.

제안한 모델은 
- LR 이 크거나 warmup 이 너무 짧을 때 collapse 가 발생함. 
- EMA decay rate 가 낮을 때, student 모델이 collapse 되고 이게 teacher 에 다시 영향을 미침. **τ0, τe 및 τn 이 중요한 param** 
- adjacent target이 상관관계가 큼. speech 와 같이 긴 범위를 masking 해야 할 때 collapse.  
- 이를 위해 VICReg 처럼 explicit loss 를 쓰고, target representation 을 normalize 해 봄.
- VICReg 는 작은 모델에서는 잘 작동, 더 큰 모델에서는 덜 안정적이며 추가로 hyperparam을 조정해야 함. 
- 근데 normalization 만 해도 충분하더라.
